client/circuitbreaker/circuit_breaker.go

```diff
@@ -0,0 +1,280 @@
+	cb.successCounter = m.CircuitBreakerCounters.WithLabelValues(metricName, "success")
+	cb.failureCounter = m.CircuitBreakerCounters.WithLabelValues(metricName, "failure")
+	cb.fastFailCounter = m.CircuitBreakerCounters.WithLabelValues(metricName, "fast_fail")
+	return cb
+}
+
+// ChangeSettings changes the CircuitBreaker settings.
+// The changes will be reflected only in the next evaluation window.
+func (cb *CircuitBreaker[T]) ChangeSettings(apply func(config *Settings)) {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	apply(cb.config)
+}
+
+// Execute calls the given function if the CircuitBreaker is closed and returns the result of execution.
+// Execute returns an error instantly if the CircuitBreaker is open.
+// https://github.com/tikv/rfcs/blob/master/text/0115-circuit-breaker.md
+func (cb *CircuitBreaker[T]) Execute(call func() (T, Overloading, error)) (T, error) {
+	state, err := cb.onRequest()
+	if err != nil {
+		var defaultValue T
+		return defaultValue, err
+	}
+
+	defer func() {
+		e := recover()
+		if e != nil {
+			cb.onResult(state, Yes)
+			panic(e)
+		}
+	}()
+
+	result, overloaded, err := call()
+	cb.onResult(state, overloaded)
+	return result, err
+}
+
+func (cb *CircuitBreaker[T]) onRequest() (*State[T], error) {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	state, err := cb.state.onRequest(cb)
+	cb.state = state
+	return state, err
+}
+
+func (cb *CircuitBreaker[T]) onResult(state *State[T], open Overloading) {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	if cb.state == state {
+		state.onResult(open)
+	} // else the state moved forward so we don't need to update the counts
+}
+
+// State represents the state of CircuitBreaker.
+type State[T any] struct {
+	stateType StateType
+	cb        *CircuitBreaker[T]
+	end       time.Time
+
+	pendingCount uint32
+	successCount uint32
+	failureCount uint32
+}
+
+// newState creates a new State with the given configuration and reset all success/failure counters.
+func (cb *CircuitBreaker[T]) newState(now time.Time, stateType StateType) *State[T] {
+	var end time.Time
+	var pendingCount uint32
+	switch stateType {
+	case StateClosed:
+		end = now.Add(cb.config.ErrorRateWindow)
+	case StateOpen:
+		end = now.Add(cb.config.CoolDownInterval)
+	case StateHalfOpen:
+		// we transition to HalfOpen state on the first request after the cooldown period,
+		// so we start with 1 pending request
+		pendingCount = 1
+	default:
+		panic("unknown state")
+	}
+	return &State[T]{
+		cb:           cb,
+		stateType:    stateType,
+		pendingCount: pendingCount,
+		end:          end,
+	}
+}
+
+// onRequest transitions the state to the next state based on the current state and the previous requests results
+// All state transitions happens at the request evaluation time only
+// The implementation represents a state machine effectively
+func (s *State[T]) onRequest(cb *CircuitBreaker[T]) (*State[T], error) {
+	var now = time.Now()
+	switch s.stateType {
+	case StateClosed:
+		if s.end.Before(now) {
```

Comment:
> Each state is bound by time. When circuit breaker moves to a new state, it defines sets the deadline (end time) for this state.

Yes, but I mean during one state, we only need to check it once. For example:

- After `StateHalfOpen` switches to `StateClosed`. Before it switch to next state, we only need to check it once. Once the condition `s.end.Before(now)` is met, we do not need check it again, because after that time, each request will meet it?

Reply:
Maybe the root problem is https://github.com/tikv/pd/pull/8856/files/a95b198e3f3ba8355def44b1210237d67c9e9a19#r1865433614 mentioned.

Reply:
This comment has been resolved by https://github.com/tikv/pd/pull/8856#discussion_r1875457345